feat: perf.rs - token counting analysis

ryanolson opened the pull request:
## Summary

Adds comprehensive token counting analysis infrastructure to Dynamo for LLM streaming responses and logprob analysis.

**🔄 Updated: Successfully merged main branch (32 commits) preserving all token counting features**

## Key Features

### Token Counting Analysis
- Stream analysis with per-choice token counting and validation
- Error detection (tokens after finish, missing choices, incomplete streams)
- Optional tokenizer integration for accurate counts
- Token generation rate calculations

### Logprob Analysis  
- Greedy vs non-greedy decoding pattern detection
- Token selection confidence metrics using logprob distributions
- Sensitivity analysis for close probability alternatives

### GGUF Tokenizer Support
- Native GGUF model tokenizer integration
- Performance optimized with ahash
- Supports Llama, GPT-2, and other GGUF models

## Merge Highlights

✅ **Main Branch Integration (32 commits)**
- Integrated ModelExpress client, prefix matcher utilities, CUDA fatbin support
- Enhanced backend health checks and monitoring
- Upgraded to Rust 1.90 with clean clippy/fmt
- Fixed blake3 API deprecation and utils module conflicts
- All 323 tests passing

## Usage

```rust
// Token analysis
let analysis = analyze_token_counting(recorded_stream, Some(&tokenizer), "request-id")?;
println!("Tokens: {}, Rate: {:.2}/sec", analysis.total_tokens, analysis.token_rate);

// Logprob analysis  
let analysis = analyze_logprobs(recorded_stream, 0.1)?;
println!("Greedy: {}, Confidence: {:.3}", analysis.greedy_count, analysis.average_confidence);
```

## Files Changed
- `lib/llm/src/perf/tokens.rs` - Core token counting analysis
- `lib/llm/src/perf/logprobs.rs` - Logprob analysis utilities
- `lib/llm/src/gguf/` - GGUF tokenizer integration  
- `lib/llm/src/utils/gzip.rs` - Compressed model utilities
- Test coverage with 323 passing tests

Ready for production use with full main branch compatibility.

ryanolson committed:
adding byot feature

ryanolson committed:
async-openai features added at top-level; removing features from the wrong package

ryanolson committed:
updates

ryanolson committed:
refactor

ryanolson committed:
reducing scope on recorded stream

ryanolson committed:
fix: resolve merge conflicts in perf module and http service tests

Resolved conflicts between performance recording module implementation and
empty file, keeping the full implementation. Also resolved conflicts in
HTTP service tests to use proper API base configuration and service
readiness checking.

-Agent Generated Commit Message

ryanolson committed:
full end-to-end tests with captures vllm sse streams

ryanolson committed:
adding capture notes

ryanolson committed:
big change: dropping Sync requirement from Resp in AsyncEngine<Req, Resp, Error>

ryanolson committed:
rebased off main - ugly mismatch from ryan/http-client

ryanolson committed:
removed comments

ryanolson committed:
Merge branch 'main' into ryan/logprob-v2

ryanolson committed:
missing test data file

ryanolson committed:
Merge branch 'ryan/logprob-v2' of github.com:ai-dynamo/dynamo into ryan/logprob-v2

ryanolson committed:
Merge branch 'main' into ryan/logprob-v2

ryanolson committed:
update filename

ryanolson committed:
Merge branch 'main' into ryan/logprob-v2

ryanolson committed:
updates to lobprobs api and data structures

ryanolson committed:
Merge branch 'ryan/logprob-v2' of github.com:ai-dynamo/dynamo into ryan/logprob-v2

ryanolson committed:
perform token counting analysis

ryanolson committed:
updates

ryanolson committed:
merge main

ryanolson committed:
update deny.toml

ryanolson committed:
Merge branch 'main' into ryan/token-counting-analysis

grahamking commented on lib/llm/src/perf.rs:
```diff
@@ -340,6 +348,61 @@ pub fn record_response_stream<R: Data + Clone>(
+) -> Result<DataStream<Annotated<T>>, anyhow::Error> {
+    // Read the entire file as a string
+    let data =
+        std::fs::read_to_string(path).map_err(|e| anyhow::anyhow!("Failed to read file: {}", e))?;
```
Can you put `path` in the error? Otherwise we get the annoying UNIX "No such file or directory" but it doesn't say which one.

ryanolson replied:
done

ryanolson committed:
update deny.toml

ryanolson committed:
Merge branch 'ryan/token-counting-analysis' of github.com:ai-dynamo/dynamo into ryan/token-counting-analysis

ryanolson committed:
adding filename/path to the read file error

ryanolson committed:
Merge branch 'main' into ryan/token-counting-analysis

ryanolson committed:
adding libbz2

ryanolson committed:
libbz2 -> bzip2

ryanolson committed:
bzip2 -> libbz2-1.0

ryanolson committed:
updates to reduce box pinning streams

ryanolson committed:
remove ci change

ryanolson committed:
trying libbz2-dev

ryanolson committed:
pivot: bzip2 -> gzip

ryanolson committed:
adding git-lfs

ryanolson committed:
adding checkotu with lfs

ryanolson committed:
try to improve ci caching

ryanolson committed:
Merge branch 'main' into ryan/token-counting-analysis

ryanolson committed:
update ci for caching

ryanolson committed:
updating ci again

ryanolson committed:
Merge branch 'ryan/token-counting-analysis' of github.com:ai-dynamo/dynamo into ryan/token-counting-analysis

ryanolson committed:
still seeing slow ci; merging slow cargo bits together

ryanolson committed:
more ci

ryanolson committed:
descoping

ryanolson committed:
clean up updated ci file

ryanolson committed:
adding a ci profile

ryanolson committed:
invalidate cache

ryanolson committed:
adding ci profiles everywhere

ryanolson committed:
missing ci profile

ryanolson committed:
inherits is mandatory

ryanolson committed:
update gguf tokenizer to address upstream changes

ryanolson committed:
Merge branch 'main' into ryan/token-counting-analysis

ryanolson committed:
not sure why but tests in the vllm container are failing

ryanolson committed:
Merge branch 'ryan/token-counting-analysis' of github.com:ai-dynamo/dynamo into ryan/token-counting-analysis

ryanolson commented:
WIP

ryanolson committed:
refactor: remove deepseek tokenizer and replace with mock tokenizer

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
feat: merge main branch and resolve conflicts

Resolved merge conflicts while preserving token counting improvements:
- Updated dependencies to latest versions from main
- Kept gzip compression utilities and token counting features
- Resolved import conflicts in logprobs.rs and tokens.rs
- Updated workflow files with latest CI improvements
- Removed deprecated tools.rs file as deleted in main
- Regenerated Cargo.lock files with updated dependencies
- Fixed API changes in blake3 and response structures

Token counting analysis features preserved:
- Token counting infrastructure with multiple data sources
- Performance recording framework for streaming responses
- GzipExtractor utility for compressed file handling
- Mock tokenizer tests replacing deepseek dependency

All tests pass successfully after merge.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson commented:
/ok to test

ryanolson committed:
feat: merge main branch into token-counting-analysis

Merged main branch resolving conflicts and preserving token counting features.

**Conflict Resolutions:**
- .gitattributes: Added *.fatbin binary handling
- .gitignore: Preserved SDK binaries ignore entry
- Cargo.toml: Integrated ModelExpress dependencies, used main's rand 0.9.0
- lib/llm/Cargo.toml: Kept flate2/ahash deps, accepted ModelExpress integration
- Cargo.lock: Regenerated for dependency resolution

**Code Fixes:**
- Fixed blake3 API: Replaced deprecated mmap methods with std::fs::read
- Fixed response struct: Removed .inner field access in TokenExtractor
- Fixed utils module conflict: Merged gzip and prefix_matcher modules
- Fixed test imports: Updated to dynamo_async_openai
- Fixed clippy warnings: Collapsed nested if, removed unnecessary to_string()
- Added missing reasoning_content field to ChatCompletionStreamResponseDelta

All tests passing (323), clippy clean, formatted with Rust 1.90.

ryanolson committed:
chore: remove old utils.rs file after module restructure

ryanolson committed:
style: format code with Rust 1.90 rustfmt
